File: src/sessions.ts
import type { Workout } from "./models.ts";
import { calendarDay } from "./models.ts";

export interface Session {
  date: string; // "2026-03-07"
  workouts: Workout[];
  totalDistance: number;
  totalTime: number; // tenths of seconds
}

export function groupIntoSessions(workouts: Workout[]): Session[] {
  const byDay = new Map<string, Workout[]>();

  for (const w of workouts) {
    const day = calendarDay(w);
    const existing = byDay.get(day);
    if (existing) {
      existing.push(w);
    } else {
      byDay.set(day, [w]);
    }
  }

  return Array.from(byDay.entries())
    .map(([date, ws]) => ({
      date,
      workouts: ws.sort((a, b) => a.date.localeCompare(b.date)),
      totalDistance: ws.reduce((sum, w) => sum + w.distance, 0),
      totalTime: ws.reduce((sum, w) => sum + w.time, 0),
    }))
    .sort((a, b) => a.date.localeCompare(b.date));
}

/** Count unique calendar days (sessions) in a list of workouts. */
export function sessionCount(workouts: Workout[]): number {
  const days = new Set(workouts.map(calendarDay));
  return days.size;
}

File: internal/stats/sessions.go
package stats

import (
	"sort"

	"github.com/richhaase/c2/internal/model"
)

type Session struct {
	Date          string
	Workouts      []model.Workout
	TotalDistance int
	TotalTime     int
}

func GroupIntoSessions(workouts []model.Workout) []Session {
	byDay := make(map[string][]model.Workout)
	for _, workout := range workouts {
		day := model.CalendarDay(workout)
		byDay[day] = append(byDay[day], workout)
	}

	sessions := make([]Session, 0, len(byDay))
	for day, dayWorkouts := range byDay {
		sort.Slice(dayWorkouts, func(i, j int) bool {
			return dayWorkouts[i].Date < dayWorkouts[j].Date
		})

		session := Session{
			Date:     day,
			Workouts: dayWorkouts,
		}
		for _, workout := range dayWorkouts {
			session.TotalDistance += workout.Distance
			session.TotalTime += workout.Time
		}
		sessions = append(sessions, session)
	}

	sort.Slice(sessions, func(i, j int) bool {
		return sessions[i].Date < sessions[j].Date
	})
	return sessions
}

func SessionCount(workouts []model.Workout) int {
	days := make(map[string]struct{})
	for _, workout := range workouts {
		days[model.CalendarDay(workout)] = struct{}{}
	}
	return len(days)
}
